### ISSUE TITLE
Git preferences

### ISSUE THREAD

**Issue by SimonBiggs** [pymedphys/pymedphys#14]
In GitLab by @SimonBiggs on Dec 6, 2018, 07:38

Hi @Centrus007,

What are your preferences regarding me commiting to a branch you are currently working on? Would you prefer I directly commit to that branch, or would it be better me branching off of the branch you're working on then creating a merge request on GitLab so that you have a chance to see it/discuss it on GitLab before it is merged into your wip feature branch.

Cheers,
Simon

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 6, 2018, 07:39

changed the description

**Comment by SimonBiggs** [comment]
In GitLab by @Centrus007 on Dec 6, 2018, 12:30

It's a tricky one, because I think this really comes down to task allocation, which isn't so easy for open source collaborations! 

The answer I'd give would depend on the case at hand. But in general, my preference is to [**scope branches according to what a *single* developer is working on**](https://stackoverflow.com/questions/14865283/proper-git-workflow-scheme-with-multiple-developers-working-on-same-task). This means compartmentalising - and often narrowing - the scope of tasks. But if your work cannot be compartmentalised - i.e. it's a clear, serial progression from previous commits - work on the same branch. Especially if it will only constitute a small number of commits.

In general I'd say it's better to have developers work on parallel tasks, with code reviews after tentative task completion. This does **not** preclude communication between developers throughout development - this is still **strongly** encouraged.

For example: we will both be working on the dicom anonymiser. Generally, I think it's more efficient for us to work within our own branches on compartmentalised portions of the dicom anonymiser. Compartments might include the anonymisation function itself, creating a CLI for it, creating a test function for it, etc. These can happen in parallel in separate branches. In this model, each developer would pick - or in a typical software development team *be assigned* - to a compartment. Developers might make small changes/individual commits on another branch, but I'd suggest long pieces of work on the same task is better avoided (there's plenty more stuff to contribute!)

So we could narrow the scope of *14-anonymise-dicom* to include only the anonymisation function itself (I don't think it makes sense to narrow further than that). CLI, testing and maybe even documentation could constitute separate compartments with their own branches. But small bugfixes, mild code refactoring, etc. probably do not deserve their own branches.

How does that sound? Does it make sense?

As an aside, this is where [git flow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) (or perhaps [GitLab flow](https://docs.gitlab.com/ee/workflow/gitlab_flow.html)) shines! Once you have multiple developers contributing to less distinct portions of the project, having a dev branch really does help with maintaining the latest development code.

**Comment by SimonBiggs** [comment]
In GitLab by @Centrus007 on Dec 6, 2018, 12:36

After all of that, I'm not even sure I answered your direct question.

It depends on what you want to do. If you're happy enough with the dicom anonymiser function, I'd ordinarily merge that branch to *master* (actually, *develop*, but we're not Git flowing) - not necessarily close it though - and then you could branch from *master* to add, say, CLI. But if you want to add a few small features directly to the function, or fix a bug, or refactor code etc. before merging, just commit to the existing branch.

Those are my two cents, anyway.

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 6, 2018, 13:20

Yup brilliant. Makes sense. I like that process. I'm not completely against
git flow / gitlab flow. If you think it would help our process I am all for
trying it out.

If anything, right now is the ideal time, while it's only us three
contributing, to be experimenting with the git workflow.

**Comment by SimonBiggs** [comment]
In GitLab by @SimonBiggs on Dec 14, 2018, 22:01

closed
